ext/cache/lib.rs

```diff
@@ -0,0 +1,207 @@
+  pub request_url: String,
+}
+
+#[async_trait]
+pub trait Cache: Clone {
+  async fn storage_open(&self, cache_name: String) -> Result<i64, AnyError>;
+  async fn storage_has(&self, cache_name: String) -> Result<bool, AnyError>;
+  async fn storage_delete(&self, cache_name: String) -> Result<bool, AnyError>;
+
+  async fn put(
+    &self,
+    request_response: CachePutRequest,
+  ) -> Result<Option<Rc<dyn Resource>>, AnyError>;
+  async fn r#match(
+    &self,
+    request: CacheMatchRequest,
+  ) -> Result<
+    Option<(CacheMatchResponseMeta, Option<Rc<dyn Resource>>)>,
+    AnyError,
+  >;
+  async fn delete(&self, request: CacheDeleteRequest)
+    -> Result<bool, AnyError>;
+}
+
+#[op]
+pub async fn op_cache_storage_open<CA>(
+  state: Rc<RefCell<OpState>>,
+  cache_name: String,
+) -> Result<i64, AnyError>
+where
+  CA: Cache + 'static,
+{
+  let cache = get_cache::<CA>(&state)?;
+  cache.storage_open(cache_name).await
+}
+
+#[op]
+pub async fn op_cache_storage_has<CA>(
+  state: Rc<RefCell<OpState>>,
+  cache_name: String,
+) -> Result<bool, AnyError>
+where
+  CA: Cache + 'static,
+{
+  let cache = get_cache::<CA>(&state)?;
+  cache.storage_has(cache_name).await
+}
+
+#[op]
+pub async fn op_cache_storage_delete<CA>(
+  state: Rc<RefCell<OpState>>,
+  cache_name: String,
+) -> Result<bool, AnyError>
+where
+  CA: Cache + 'static,
+{
+  let cache = get_cache::<CA>(&state)?;
+  cache.storage_delete(cache_name).await
+}
+
+#[op]
+pub async fn op_cache_put<CA>(
+  state: Rc<RefCell<OpState>>,
+  request_response: CachePutRequest,
+) -> Result<Option<ResourceId>, AnyError>
+where
+  CA: Cache + 'static,
+{
+  let cache = get_cache::<CA>(&state)?;
+  match cache.put(request_response).await? {
+    Some(resource) => {
+      let rid = state.borrow_mut().resource_table.add_rc_dyn(resource);
+      Ok(Some(rid))
+    }
+    None => Ok(None),
+  }
+}
+
+#[op]
+pub async fn op_cache_match<CA>(
+  state: Rc<RefCell<OpState>>,
+  request: CacheMatchRequest,
+) -> Result<Option<CacheMatchResponse>, AnyError>
+where
+  CA: Cache + 'static,
+{
+  let cache = get_cache::<CA>(&state)?;
+  match cache.r#match(request).await? {
+    Some((meta, None)) => Ok(Some(CacheMatchResponse(meta, None))),
+    Some((meta, Some(resource))) => {
+      let rid = state.borrow_mut().resource_table.add_rc_dyn(resource);
+      Ok(Some(CacheMatchResponse(meta, Some(rid))))
+    }
+    None => Ok(None),
+  }
+}
+
+#[op]
+pub async fn op_cache_delete<CA>(
+  state: Rc<RefCell<OpState>>,
+  request: CacheDeleteRequest,
+) -> Result<bool, AnyError>
+where
+  CA: Cache + 'static,
+{
+  let cache = get_cache::<CA>(&state)?;
+  cache.delete(request).await
+}
+
+pub fn get_cache<CA>(state: &Rc<RefCell<OpState>>) -> Result<CA, AnyError>
+where
+  CA: Cache + 'static,
+{
+  let state = state.borrow();
+  let cache = state.try_borrow::<CA>().ok_or_else(|| {
+    custom_error(
+      "NotSupported",
+      "Cache API is not supported in this context.",
+    )
+  })?;
```

Comment:
I don't think this should be a `Deno.errors.NotSupported` error. Should be a `DOMExceptionNotSupportedError` instead.